scripts/adlsgen2setup.py

```diff
@@ -110,15 +113,110 @@ async def run(self):
+        for root, dirs, files in os.walk(src_filepath):
+            
+            #iterate through the files currently obtained by os.walk() and
+            #create the filepath string for that file and add it to the filepath_list list
+            for file in files:
+                #Checks to see if the root is '.' and changes it to the correct current
+                #working directory by calling os.getcwd(). Otherwise root_path will just be the root variable value.
+                if root == '.':
+                    root_path = os.getcwd() + "/"
+                else:
+                    root_path = root
+                
+                filepath = os.path.join(root_path, file)
+                
+                #Appends filepath to filepath_list if filepath does not currently exist in filepath_list
+                if filepath not in filepath_list:
+                    filepath_list.append(filepath)
+                    
+        #Return filepath_list        
+        return filepath_list
+
+    async def scan_and_upload_directories(self, directories: dict[str, DataLakeDirectoryClient], filesystem_client):
+        logger.info("Scanning and uploading files from directories recursively...")
+        for directory, directory_client in directories.items():
+            directory_path = os.path.join(self.data_directory, directory)
+            
+             # Überprüfen, ob 'scandir' existiert und auf False gesetzt ist
+            if not self.data_access_control_format["directories"][directory].get("scandir", True):
+                logger.info(f"Skipping directory {directory} as 'scandir' is set to False")
+                continue
+            
+            groups = self.data_access_control_format["directories"][directory].get("groups", [])
+
+            # Check if the directory exists before walking it
+            if not os.path.exists(directory_path):
+                logger.warning(f"Directory does not exist: {directory_path}")
+                continue
+            
+            # Get all file paths using the walk_files function
+            file_paths = await self.walk_files(directory_path)
+
+            # Upload each file collected
+            for file_path in file_paths:
+                await self.upload_file(directory_client, file_path)
+                logger.info(f"Uploaded {file_path} to {directory}")
+
     def create_service_client(self):
         return DataLakeServiceClient(
             account_url=f"https://{self.storage_account_name}.dfs.core.windows.net", credential=self.credentials
         )
 
-    async def upload_file(self, directory_client: DataLakeDirectoryClient, file_path: str):
+    async def calc_md5(self, path: str) -> str:
+        with open(path, "rb") as file:
+            return hashlib.md5(file.read()).hexdigest()
+
+    async def check_md5(self, path: str, md5_hash: str) -> bool:
+        # if filename ends in .md5 skip
+        if path.endswith(".md5"):
```

Comment:
We actually want to move the storage of the MD5 into the Blob storage metadata for our main prepdocs strategy, as the local md5 is problematic when you switch environments. Would that be possible with ADSL2? I assume it also has the ability to store metadata, given its a type of Blob storage?
What do you think of that approach, remote MD5?

Reply:
I’ve already implemented that and will include it with some additional metadata in the next commit. If iread the file directly from the source (local storage or blob/data lake storage) and create a new checksum on the fly, keep it in heap . Then, i can compare it with the persistent checksum on the target storage (datalake)and update the blob only if diff and as well the checksum in the meta.

I’ve also started on preodocs.sh to make use of the persist MD5 on the backend once it’s "injected" from the data lake using the datalake strategy which seems to be unstable for exceptions . (exits the whole loop whi is even more painful if you don't have a resume. md5 will not dupe work ok, but it still  will need to iterate again over all the offset until where the ingestion died abnormally.  For external systems i use queries with timestamps to find a proper point to resume but i have no idea if and how current datalake file walker creates any order  or if is completely random.
Maybe there is a way to walk/iterate  along a query based on a metadata field (i have added) like lastchange date in metadadata? This way i could storage the last i9nkestion date seen in another blob and use it as resume

I’m unsure where best to store this: 
Two options i had in mind

a) In the index, where it would be associated with each chunk, meaning we’d need to locate chunks where there are more than zero results for the fingerprint, updating multiple rows accordingly.

b) In /content (blob storage) for inline rendering of the source document as a citation in the browser. This has the benefit of only requiring one update per change, though we’d still need to locate and remove all matching chunks in the index by primary key for consistency.

Option (a) seems more logical, doesn’t it?

Btw i remember i had trouble with md5 dupes for bigger data amount , so maybe its better to use sha digests?

And even more important: how would above efforts fit into integrated-vectorization?
Will that work with whatever fingerprint being takeover into the index. I assume it requires a new skill, doesn't it?

Reply:
Blob storage has embedded md5 already as it seems -

`
from azure.storage.blob import BlobServiceClient

# Your connection string
connection_string = "your_connection_string"

# Your blob container name and blob name
container_name = "your_container_name"
blob_name = "your_blob_name"

# Initialize a BlobServiceClient
blob_service_client = BlobServiceClient.from_connection_string(connection_string)

# Get a reference to the blob
blob_client = blob_service_client.get_blob_client(container=container_name, blob=blob_name)

# Get the properties of the blob
blob_properties = blob_client.get_blob_properties()

# Retrieve the MD5 hash value from the properties
md5_hash = blob_properties.content_settings.content_md5

print(f"MD5 Hash of the blob: {md5_hash}")
`

If that md5 will have same value as a local computed md5 from file is the question
Also as long  as we use as blob name some filename (instead of md5) we still need to rely on that and can't handle rename scenarios. Also we can't  easily with O(1) find out if we have any blob up there for the md5 of the file in our hands. If we would use md5 as blob name it would make things in code much slimmer, but debugging and using the azure portal ui file browser for debugging a bit more cumberstone

Reply:
Blob service does have its own md5, but it only computes it for small files, so we would need to compute our own hash. Using sha256 also seems fine if we can store that in the blob metadata.